Docs: URLSearchParams (#63)
trilitech/jstz: docs/api/url_search_params.md

# 🔗 URLSearchParams

`jstz`'s implementation of the `URLSearchParams` API defines utility methods for working with the query string of a URL according to the [URL specification](https://url.spec.whatwg.org/#urlsearchparams). It is used for building and manipulating search parameters.

## Example

```typescript
// Parse query string from URL
let url = new URL("https://example.com?foo=1&bar=2");
let params = new UrlSearchParams(url.search);

// Add a new parameter
params.append("baz", 3);

// Remove parameter
params.delete("bar");
```

## Constructor

### `new UrlSearchParams(init?: [string, string][] | Record<string, string> | string)`

Creates a new instance of `UrlSearchParams` with the provided key-value pairs.

- **init**:
  One of:
  - An array of key-value pairs. Each pair is an array where the first element is the key (Name) and the second is the value.
  - A record of string keys and string values.
  - A string, which will be parsed from application/x-www-form-urlencoded format. A leading '?' character is ignored.

## Instance Properties

> ### `size: number`

Reviewer: ```suggestion
### `readonly size: number`
```

# 🔗 URLSearchParams

> `jstz` implementation of an API with utility methods for working with the query string of a URL according to the [URLSearchParams specification](https://url.spec.whatwg.org/#urlsearchparams). It is used for building and manipulating search parameters.

Reviewer: `jstz` -> `jstz`'s
 ```suggestion
`jstz`'s implementation of the `URLSearchParams` API defines utility methods for working with the query string of a URL according to the [URL specification](https://url.spec.whatwg.org/#urlsearchparams). It is used for building and manipulating search parameters.
```

## Example

```typescript
// Parse query string from URL
let url = new URL("https://example.com?foo=1&bar=2");
let params = new UrlSearchParams(url.search);

// Add a new parameter
params.append("baz", 3);

// Remove parameter
params.delete("bar");

// Loop through parameters
for (let [key, value] of params) {
  // ...

> }

Reviewer: We don't yet support an iterator on `URLSearchParams`

```

## Constructor

> ### `UrlSearchParams(values: Array<[Name: string, Value: string]>, url?: URL)`

Reviewer: ```suggestion
### `new UrlSearchParams(init?: [string, string][] | Record<string, string> | string)`
```

In future always refer to the `jstz-types` package for the types + parameter names

Creates a new instance of `UrlSearchParams` with the provided key-value pairs.

- **values**: An array of key-value pairs. Each pair is an array where the first element is the key (Name) and the second is the value.

## Instance Methods

### `setValues(values: Array<[Name: string, Value: string]>): void`

Sets the current values with the provided array of key-value pairs.

- **values**: An array of key-value pairs. Each pair consists of a name (key) and value.

### `setUrl(url: JsNativeObject<Url>): void`

Associates the `UrlSearchParams` with a given `Url` object.

- **url**: A reference to the `Url` object.

### `len(): number`

> - **Returns**: The number of search parameters present.

Reviewer: These methods don't exist. I think you're documenting the _Rust_ interface here, but we want to document the _JavaScript_ interface

### `append(name: string, value: string): void`

Appends a specified key/value pair as a new search parameter.

- **name**: The name of the search parameter.

> - **value**: The value of the search parameter.

Reviewer: We don't use this style of specifying arguments in our documentation. I'm not opposed to it, but it is inconsistent with what the rest of the docs does

> ### `remove(name: string, value?: string): void`

Reviewer: remove->delete

Removes search parameters that match the given name. If a value is provided, only parameters with that name-value pair are removed.

- **name**: The name of the search parameter to be removed.
- **value** (optional): The specific value of the search parameter to be removed.

### `get(name: string): string | null`

Returns the first value associated with the given search parameter.

- **name**: The name of the search parameter.

- **Returns**: The value associated with the given search parameter or `null` if not found.

> ### `getAll(name: string): Array<string>`

Reviewer: `Array<string>` -> `string[]`

Returns all the values associated with a given search parameter.

- **name**: The name of the search parameter.

- **Returns**: An array of values associated with the given search parameter.

> ### `contains(name: string, value?: string): boolean`

Reviewer: `contains` -> `has`
